### PR TITLE
Make cursors thread safe

### PR DESCRIPTION
**Pull request by meiravgri**
In this PR we re-organized the cursors management method

 global cursors list 
* name was changed to `g_CursorsList` in order to emphasize that it is global.
* The cursors list no longer holds  a spec info dict - instead, **each cursor holds a weak reference** to its associated spec (if exists) that needs to be promoted to a strong reference upon execution (aka, reading from the cursor). **If the promotion fails,** it means that the spec in no longer valid and the cursor needs to be released.
**In a coordinator ctx**, for a `WITHCURSOR` query, we start the cursor with a dummy spec ref that points to NULL.
In the shards, **at the cursor's first initialization,** it is assumed to be called with a valid strong ref, so there is no need to promote it by the cursor.
* The index is no longer responsible to free cursors when it is dropped. The cursors are released when the cursors GC function is called or when the server is down.

spec changes:
* the spec holds a pointer to its own ref for quick access.
* statics related to the spec's cursors were moved to the spec struct. To update them the cursor has to take strong reference to the spec.

### CODE REVIEW

File: src/aggregate/aggregate_exec.c [hunk 1]
```diff
@@ -739,20 +738,36 @@ static void runCursor(RedisModuleCtx *outputCtx, Cursor *cursor, size_t num) {
...
   if (cursor == NULL) {
     RedisModule_ReplyWithError(ctx, "Cursor not found");
     return;
   }
-  // TODO: verify spec strong ref, and release it at the end
   QueryError status = {0};
+  StrongRef execution_ref = {0};
+  bool has_spec = cursor->spec_ref.rm != NULL;
+  // If the cursor is associated with a spec
+  if(has_spec) {
```
Comment (DvirDukhan) [inline src/aggregate/aggregate_exec.c:750]: what is the opposite case? when a cursor does not associated with a spec?

File: src/spec.h [hunk 2]
```diff
@@ -305,6 +305,13 @@ typedef struct IndexSpec {
   // Current spec version.
   // Should be updated after acquiring the write lock.
   size_t specVersion;
+
+  // Cursors counters
+  size_t cursorsCap;
+  size_t activeCursors;
+
+  // Quick access to the spec's strong ref
+  StrongRef own_ref;
```
Comment (DvirDukhan) [inline src/spec.h:314]: is this mandatory?

File: src/cursor.c [hunk 3]
```diff
@@ -142,26 +152,12 @@ int Cursors_CollectIdle(CursorList *cl) {
...
-}
-
-void CursorList_RemoveSpec(CursorList *cl, const char *k) {
-  CursorSpecInfo *info = findInfo(cl, k);
-  if (info) {
-    dictDelete(cl->specsDict, k);
-    rm_free(info->keyName);
-    rm_free(info);
-  }
+void Cursors_initSpec(IndexSpec *spec, size_t capacity) {
```
Comment (DvirDukhan) [inline src/cursor.c:155]: I think this should be a part of the spec creation and not related to the cursor h/c files

File: src/cursor.c [hunk 4]
```diff
@@ -180,54 +176,54 @@ static uint64_t CursorList_GenerateId(CursorList *curlist) {
...
   Cursor *cur = NULL;
 
-  if (spec == NULL) {
-    QueryError_SetErrorFmt(status, QUERY_ENOINDEX, "Index `%s` does not have cursors enabled",
-                           lookupName);
-    goto done;
-  }
+  // If the cursor should be associated with a spec, 
+  // we assume that global_spec_ref points to a valid spec, else the function returns NULL.
+  IndexSpec *spec = StrongRef_Get(global_spec_ref);
```
Comment (DvirDukhan) [inline src/cursor.c:187]: please have this function built as
```c
if(!spec) {
    // Do what is need to be done without a spec
} else {
    // Do with spec
}
```

File: src/cursor.c [hunk 5]
```diff
@@ -180,54 +176,54 @@ static uint64_t CursorList_GenerateId(CursorList *curlist) {
...
 
-  if (spec->used >= spec->cap) {
+  // If we are in a coordinator ctx, the spec is NULL
+  if (spec && (spec->activeCursors >= spec->cursorsCap)) {
     /** Collect idle cursors now */
     Cursors_GCInternal(cl, 0);
-    if (spec->used >= spec->cap) {
+    if (spec->activeCursors >= spec->cursorsCap) {
```
Comment (DvirDukhan) [inline src/cursor.c:193]: you need to demote strong ref here

File: src/cursor.c [hunk 6]
```diff
@@ -283,12 +279,11 @@ int Cursors_Purge(CursorList *cl, uint64_t cid) {
 }
 
 int Cursor_Free(Cursor *cur) {
-  return Cursors_Purge(cur->parent, cur->id);
+  return Cursors_Purge(&g_CursorsList, cur->id);
 }
 
-void Cursors_RenderStats(CursorList *cl, const char *name, RedisModuleCtx *ctx) {
+void Cursors_RenderStats(CursorList *cl, IndexSpec *spec, RedisModuleCtx *ctx) {
```
Comment (DvirDukhan) [inline src/cursor.c:285]: assuming spec is locked?

File: tests/pytests/test_cursors.py [hunk 7]
```diff
@@ -148,3 +148,23 @@ def testLeaked(env):
     # Test ensures in CursorList_Destroy() checks shutdown with remaining cursors
     loadDocs(env)
     env.expect('FT.AGGREGATE idx * LOAD 1 @f1 WITHCURSOR COUNT 1 MAXIDLE 1')
+
+def testIndexDropWhileIdle(env):
```
Comment (DvirDukhan) [inline tests/pytests/test_cursors.py:152]: not going to background?

File: src/cursor.c [hunk 8]
```diff
@@ -180,54 +176,54 @@ static uint64_t CursorList_GenerateId(CursorList *curlist) {
...
   }
 
   cur = rm_calloc(1, sizeof(*cur));
-  cur->parent = cl;
-  cur->specInfo = spec;
   cur->id = CursorList_GenerateId(cl);
   cur->pos = -1;
   cur->timeoutIntervalMs = interval;
+  if(spec) {
+    cur->spec_ref = StrongRef_Demote(global_spec_ref);
```
Comment (meiravgri) [inline src/cursor.c:205]: switch lines

File: src/cursor.c (hunk 5 above)
Comment (meiravgri) [reply src/cursor.c:193]: Note that this function gets a strong refernce, so it is assumed that the management of the reference is done by the calling function.

File: tests/pytests/test_cursors.py (hunk 7 above)
Comment (meiravgri) [reply tests/pytests/test_cursors.py:152]: will add

File: src/cursor.c (hunk 8 above)
Comment (meiravgri) [inline src/cursor.c:205]: add comment

File: src/cursor.c [hunk 9]
```diff
@@ -51,26 +47,38 @@ static void Cursor_RemoveFromIdle(Cursor *cur) {
...
     cur->execState = NULL;
   }
+  // if There's a spec associated with the cursor
+  if(cur->spec_ref.rm) { 
+    StrongRef spec_ref = WeakRef_Promote(cur->spec_ref);
+    IndexSpec *spec = StrongRef_Get(spec_ref);
+    // the spec may have been dropped, so we need to make sure it is still valid. 
+    if(spec) {
+      spec->activeCursors--;
+      StrongRef_Release(spec_ref);
```
Comment (oshadmi) [inline src/cursor.c:77]: The cursor might now keep the index alive for up to `CURSOR_MAX_IDLE` ?

Comment (meiravgri) [reply src/cursor.c:77]: No the strong reference is released after each read
We keep the weak ref in the cursor's struct to validate that the spec still exists when we try to call read. If the weak ref promotion fails, the query fails

File: src/spec.c [hunk 10]
```diff
@@ -1059,6 +1059,7 @@ int IndexSpec_AddFields(StrongRef spec_ref, IndexSpec *sp, RedisModuleCtx *ctx,
 StrongRef IndexSpec_Parse(const char *name, const char **argv, int argc, QueryError *status) {
   IndexSpec *spec = NewIndexSpec(name);
   StrongRef spec_ref = StrongRef_New(spec, (RefManager_Free)IndexSpec_Free);
+  spec->own_ref = spec_ref;
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/spec.c:1071]: No this is not allocated, just copied from the strong ref the spec is allocationg
The strong one ref ia released when the spec is dropped

[the remark this answers is not in the document]
Comment (oshadmi) [reply src/spec.c:1071]: I meant do we need to set it to NULL etc., so it cannot be used even during freeCB `IndexSpec_Free`,
but I suppose any freeing code should just not use it.

[the remark this answers is not in the document]
Comment (meiravgri) [reply src/spec.c:1071]: I wrapped it in a function and added and nullify in IndexSpec_Free

**Review by oshadmi — APPROVED** [review]

**Review by DvirDukhan — APPROVED** [review]

**Pull request merged**